### PR TITLE
enable passing server options for notebook server launch

### LINKED ISSUE #65: include possibility to add GPU requirements to pod manifest
Add options to include a GPU request in the pod manifest.

### LINKED ISSUE #66: use server options for launching jupyter servers
- [ ] provide a `server_options` endpoint
- [ ] consume `server_options` from `POST` body and relay to the spawner
- [ ] modify the spawner to include the server options in the pod manifest

### PR DESCRIPTION
**Pull request by rokroskar**
This PR allows the notebooks service setup to include default server options -- and to override them upon server launch. The options are set in `values.yaml` and written into a file in the container via a `configMap`. 

The options can be queried by a UI client using `GET <service_prefix>/<namespace>/<project>/<commit-sha>/server_options` -- the returned JSON can be used to render a template. 

The client should return only the option and the value, e.g. if the JSON server from `server_options` is

```
{
  "serverOptions":
  { 
    "resources":
      { 
        "cpu_limit":
         {
          "default": 1,
          "type": int,
          "range": [1,4]
          }
      }
  }
}
```

The client should use 

```
{
  "serverOptions":
    {
      "resources":
      {
         "cpu_limit": "1"
      }
    }
}
```

as the json data in the POST request to spawn the server. 
  
---

closes #66 
closes #65 
addresses https://github.com/SwissDataScienceCenter/renku/issues/443

### CODE REVIEW

File: helm-chart/renku-notebooks/Chart.yaml [hunk 1]
```diff
@@ -2,4 +2,4 @@ apiVersion: v1
 appVersion: '1.0'
 description: A Helm chart for the Renku Notebooks service
 name: renku-notebooks
-version: 0.2.0
+version: 0.0-unclean-191b684
```
Comment (leafty) [inline helm-chart/renku-notebooks/Chart.yaml]: :thinking:

Comment (rokroskar) [reply helm-chart/renku-notebooks/Chart.yaml]: Yes I know, I'll clean all this up when I'm done running it on my minikube :)

File: helm-chart/renku-notebooks/templates/deployment.yaml [hunk 2]
```diff
@@ -59,11 +61,15 @@ spec:
...
             - name: IMAGE_REGISTRY
-              value: {{required "An image registry must be specified." .Values.imageRegistry }}
+              value: {{ required "An image registry must be specified." .Values.gitlab.registry.host }}
           ports:
             - name: http
               containerPort: 8000
               protocol: TCP
+          volumeMounts:
+            - name: server-options
+              mountPath: /etc/renku-notebooks/server_options.json
```
Comment (leafty) [inline helm-chart/renku-notebooks/templates/deployment.yaml]: `/etc/renku-notebooks`
(https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#add-configmap-data-to-a-volume)

File: jupyterhub/jupyterhub-k8s.Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM jupyterhub/k8s-hub:0.7.0
+FROM jupyterhub/k8s-hub:ae44b17
```
Comment (leafty) [inline jupyterhub/jupyterhub-k8s.Dockerfile]: huh?

File: jupyterhub/spawners.py [hunk 4]
```diff
@@ -285,38 +283,29 @@ def get_pod_manifest(self):
             options = self.user_options
             commit_sha_7 = options.get('commit_sha')[:7]
 
-            # https://gist.github.com/tallclair/849601a16cebeee581ef2be50c351841
-            init_container_name = 'git-clone'
```
Comment (leafty) [inline jupyterhub/spawners.py]: no name?

File: src/notebooks_service.py [hunk 5]
```diff
@@ -379,15 +380,39 @@ def launch_notebook(user, namespace, project, commit_sha, notebook=None):
...
+            'default', os.getenv('JUPYTERHUB_SINGLEUSER_DEFAULT_URL')
+        )
+    )
+    server_options.setdefault('resources', {})
+
+    for key in default_resources.keys():
+        server_options['resources'].setdefault(
+            key,
+            default_resources.get(key)['default']
+        )
```
Comment (leafty) [inline src/notebooks_service.py]: TODO: validate options

File: helm-chart/renku-notebooks/values.yaml [hunk 6]
```diff
@@ -75,14 +81,33 @@ tolerations: []
...
+
 ## Configuration for the jupyterhub service
 jupyterhub:
   rbac:
     enabled: true
   hub:
     image:
       name: renku/jupyterhub-k8s
-      tag: '0.2.0'
+      tag: '0.0-unclean-191b684'
```
Comment (leafty) [inline helm-chart/renku-notebooks/values.yaml]: :see_no_evil:

File: helm-chart/renku-notebooks/values.yaml [hunk 7]
```diff
@@ -140,14 +165,14 @@ jupyterhub:
   singleuser:
     image:
       name: renku/singleuser
-      tag: '0.2.0'
+      tag: '0.0-unclean-191b684'
```
Comment (leafty) [inline helm-chart/renku-notebooks/values.yaml]: :speak_no_evil:

File: helm-chart/renku-notebooks/values.yaml [hunk 8]
```diff
@@ -140,14 +165,14 @@ jupyterhub:
...
     storage:
       type: none
     defaultUrl: /lab
   singleuser-r:
     image:
       name: renku/singleuser-r
-      tag: 'latest'
+      tag: '0.0-unclean-191b684'
```
Comment (leafty) [inline helm-chart/renku-notebooks/values.yaml]: :hear_no_evil:

File: helm-chart/renku-notebooks/templates/deployment.yaml [hunk 9]
```diff
@@ -28,7 +28,7 @@ spec:
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
```
Comment (leafty) [inline helm-chart/renku-notebooks/templates/deployment.yaml]: add to the pod spec:
```
annotations:
  checksum/config: {{ include (print $.Template.BasePath "/configmap.yaml") . | sha256sum }}
```

File: helm-chart/renku-notebooks/values.yaml [hunk 10]
```diff
@@ -75,14 +81,35 @@ tolerations: []
...
+  resources:
+    cpu_limit:
+      displayName: Number of CPUs
+      type: float
+      default: 1.0
+      range: [0.1, 4.0]
+    mem_limit:
+      displayName: Memory
+      type: enum
+      default: 1G
```
Comment (leafty) [inline helm-chart/renku-notebooks/values.yaml]: default 2G

File: helm-chart/renku-notebooks/values.yaml [hunk 11]
```diff
@@ -75,14 +81,35 @@ tolerations: []
 
 affinity: {}
 
+serverOptions:
+  resources:
+    cpu_limit:
+      displayName: Number of CPUs
+      type: float
+      default: 1.0
+      range: [0.1, 4.0]
```
Comment (leafty) [inline helm-chart/renku-notebooks/values.yaml]: only use discrete options

File: helm-chart/renku-notebooks/values.yaml [hunk 12]
```diff
@@ -75,14 +81,35 @@ tolerations: []
...
+    mem_limit:
+      displayName: Memory
+      type: enum
+      default: 2G
+      options: [2G, 4G, 8G]
+    gpu:
+      displayName: Number of GPUs
+      type: integer
+      default: 0
+      range: [0, 4]
```
Comment (leafty) [inline helm-chart/renku-notebooks/values.yaml]: discrete options only too

File: helm-chart/renku-notebooks/values.yaml (hunk 11 above)
Comment (rokroskar) [reply helm-chart/renku-notebooks/values.yaml]: why? CPUs requests are fractional

**Comment by rokroskar** [comment]
btw, I don't pass `0` as a gpu number -- if it is `0` then we don't ask for a gpu at all.

File: helm-chart/renku-notebooks/values.yaml (hunk 11 above)
Comment (leafty) [reply helm-chart/renku-notebooks/values.yaml]: a. are you going to use it?
b. with `options: [0, 1, 2, 4, 8]` you can have `0 = no limits`
c. you can include `0.5` anyway

Comment (rokroskar) [reply helm-chart/renku-notebooks/values.yaml]: this is a limit, not a request -- so I think actually integer values are probably fine, so I propose we do `range: [1,X]` (I just pushed this change)

File: jupyterhub/spawners.py [hunk 13]
```diff
@@ -348,21 +337,51 @@ def get_pod_manifest(self):
...
+                server_options['resources'].get('cpu_guarantee', 0.1)
+            )
+            self.mem_limit = server_options['resources'].get('mem_limit')
+            self.mem_guarantee = server_options['resources'].get(
+                'mem_guarantee', '500M'
+            )
+
+            gpu = server_options['resources'].get('gpu', {})
+            if gpu:
+                self.extra_resource_guarantees = {
```
Comment (leafty) [inline jupyterhub/spawners.py]: `extra_resource_limits`

File: src/notebooks_service.py (hunk 5 above)
Comment (rokroskar) [reply src/notebooks_service.py]: Yes, I will make an issue for this -- it's not obvious to me how the validation should happen or indeed what should be validated.

**Pull request merged**